criu/sk-udp.c

```diff
@@ -0,0 +1,655 @@
+	*outq_pkt = 0;
+	return ret;
+}
+
+static int udp_lock_one(int fd, struct inet_sk_desc *sk)
+{
+	int ret;
+
+	sk->rfd = dup(fd);
+	if (sk->rfd < 0) {
+		pr_perror("Can't save socket fd for repair");
+		return -1;
+	}
+
+	if (!(root_ns_mask & CLONE_NEWNET)) {
+		ret = nf_lock_connection(sk);
+		if (ret < 0)
+			goto err;
+	}
+
+	list_add_tail(&sk->rlist, &udp_repair_sockets);
+
+	return 0;
+err:
+	close(sk->rfd);
+	return -1;
+}
+
+static void udp_unlock_one(struct inet_sk_desc *sk)
+{
+	int ret, aux = 0;
+
+	if (!(root_ns_mask & CLONE_NEWNET)) {
+		ret = nf_unlock_connection(sk);
+		if (ret < 0)
+			pr_perror("Failed to unlock UDP socket");
+	}
+
+	restore_opt(sk->rfd, SOL_UDP, UDP_REPAIR, &aux);
+	close(sk->rfd);
+}
+
+void udp_unlock(void)
+{
+	struct inet_sk_desc *sk, *n;
+
+	list_for_each_entry_safe(sk, n, &udp_repair_sockets, rlist)
+		udp_unlock_one(sk);
+}
+
+int write_img_queue(struct cr_img *img, struct udp_packet *up)
+{
+	if (pb_write_one(img, up->upe, PB_UDP_PACKET))
+		return -1;
+	if (write_img_buf(img, up->data, up->upe->pkt_len))
+		return -1;
+	return 0;
+}
+
+int dump_one_udp(int fd, struct inet_sk_desc *sk, int family)
+{
+	int aux, ret, peek_off;
+	struct cr_img *img;
+	UdpQueueEntry udp_queue_entry = UDP_QUEUE_ENTRY__INIT;
+	struct udp_packet *up, *n;
+	socklen_t len;
+
+	pr_info("Dumping a UDP socket\n");
+
+	if (dump_opt(fd, SOL_UDP, UDP_CORK, &aux))
+		return -1;
+
+	udp_queue_entry.cork = false;
+	if (aux) {
+		udp_queue_entry.cork = true;
+	}
+
+	if (udp_lock_one(fd, sk))
+		return -1;
+
+	img = open_image(CR_FD_UDP_QUEUE, O_DUMP, sk->sd.ino);
+	if (!img)
+		return -1;
+
+	ret = -1;
+	len = sizeof(peek_off);
+	if (getsockopt(fd, SOL_SOCKET, SO_PEEK_OFF, &peek_off, &len)) {
+		/* CentOS 7 does not support SO_PEEK_OFF on AF_INET sockets, the
+		 * peek off is 0 every recvmsg
+		 */
+		if (errno != EOPNOTSUPP) {
+			pr_perror("Can't read SO_PEEK_OFF");
+			goto out;
+		}
+	}
+
+	ret = dump_recv_queue(sk, family, &udp_queue_entry.inq_pkt);
+	if (ret < 0) {
+		pr_err("Dump recv queue failed\n");
+		goto restore_off;
+	}
+
+	ret = dump_send_queue(sk, family, &udp_queue_entry.outq_pkt);
+	if (ret < 0) {
+		pr_err("Dump send queue failed\n");
+		goto free_inq;
+	}
+
+	ret = pb_write_one(img, &udp_queue_entry, PB_UDP_QUEUE);
+	if (ret < 0)
+		goto free_outq;
+
+	list_for_each_entry(up, &recv_queue_packets, packet_list) {
+		ret = -1;
+		if (write_img_queue(img, up))
+			goto free_outq;
+	}
+
+	if (send_queue_packet) {
+		ret = -1;
+		if (write_img_queue(img, send_queue_packet))
+			goto free_outq;
+	}
+
+	ret = 0;
+
+free_outq:
+	if (send_queue_packet) {
+		free_udp_packet(send_queue_packet);
+		send_queue_packet = NULL;
+	}
+free_inq:
+	list_for_each_entry_safe(up, n, &recv_queue_packets, packet_list)
+		free_udp_packet(up);
+	INIT_LIST_HEAD(&recv_queue_packets);
+	xfree(g_data);
+	g_data = NULL;
+restore_off:
+	setsockopt(sk->rfd, SOL_SOCKET, SO_PEEK_OFF, &peek_off, sizeof(peek_off));
+out:
+	close_image(img);
+	return ret;
```

Comment:
pls use the exit_code pattern here.

Reply:
Can you give me an example of exit_code pattern?

Reply:
git grep exit_code ;)

The idea is that exit_code is initialized to -1 and changes to 0 only at the end. We use it in many places.

It allows to avoid bugs like this:
```
ret = -1;
if (x)
  goto err;
ret = check_smth()
if (ret)
  goto err;
if (y)
  goto err; <--- and you are in trouble...
```